internal/stores/contracts.go

```diff
@@ -179,3 +194,382 @@ func NewJSONContractStore(dir string) (*JSONContractStore, error) {
+		}
+		if err := db.AutoMigrate(tables...); err != nil {
+			return nil, err
+		}
+		if res := db.Exec("PRAGMA foreign_keys = ON", nil); res.Error != nil {
+			return nil, res.Error
+		}
+	}
+
+	return &SQLContractStore{
+		db: db,
+	}, nil
+}
+
+// AddContract implements the bus.ContractStore interface.
+func (s *SQLContractStore) AddContract(c rhpv2.Contract) error {
+	return s.db.Transaction(func(tx *gorm.DB) error {
+		fcid := c.ID()
+
+		// Prepare valid and missed outputs.
+		newValidOutputs := make([]dbValidSiacoinOutput, 0, len(c.Revision.NewValidProofOutputs))
+		for _, sco := range c.Revision.NewValidProofOutputs {
+			newValidOutputs = append(newValidOutputs, dbValidSiacoinOutput{
+				ParentID:   fcid,
+				UnlockHash: sco.UnlockHash,
+				Value:      sco.Value.Big(),
+			})
+		}
+		newMissedOutputs := make([]dbMissedSiacoinOutput, 0, len(c.Revision.NewMissedProofOutputs))
+		for _, sco := range c.Revision.NewMissedProofOutputs {
+			newMissedOutputs = append(newMissedOutputs, dbMissedSiacoinOutput{
+				ParentID:   fcid,
+				UnlockHash: sco.UnlockHash,
+				Value:      sco.Value.Big(),
+			})
+		}
+
+		// Prepare pubkeys for unlock conditions.
+		pubkeys := make([]dbSiaPublicKey, 0, len(c.Revision.UnlockConditions.PublicKeys))
+		for _, pk := range c.Revision.UnlockConditions.PublicKeys {
+			pubkeys = append(pubkeys, dbSiaPublicKey{
+				Algorithm:         pk.Algorithm,
+				Key:               pk.Key,
+				UnlockConditionID: fcid,
+			})
+		}
+
+		// Prepare unlock conditions.
+		unlockConditions := dbUnlockConditions{
+			ParentID:           fcid,
+			Timelock:           c.Revision.UnlockConditions.Timelock,
+			PublicKeys:         pubkeys,
+			SignaturesRequired: c.Revision.UnlockConditions.SignaturesRequired,
+		}
+
+		// Prepare contract revision.
+		revision := dbFileContractRevision{
+			ParentID:              fcid,
+			UnlockConditions:      unlockConditions,
+			NewRevisionNumber:     c.Revision.NewRevisionNumber,
+			NewFileSize:           c.Revision.NewFileSize,
+			NewFileMerkleRoot:     c.Revision.NewFileMerkleRoot,
+			NewWindowStart:        c.Revision.NewWindowStart,
+			NewWindowEnd:          c.Revision.NewWindowEnd,
+			NewValidProofOutputs:  newValidOutputs,
+			NewMissedProofOutputs: newMissedOutputs,
+			NewUnlockHash:         c.Revision.NewUnlockHash,
+		}
+
+		// Prepare signatures. The covered fields are stored as a blob
+		// to avoid more tables which are probably not going to be
+		// useful anyway.
+		signatures := make([]dbTransactionSignature, 0, len(c.Signatures))
+		for _, sig := range c.Signatures {
+			signatures = append(signatures, dbTransactionSignature{
+				ParentID:       fcid,
+				PublicKeyIndex: sig.PublicKeyIndex,
+				Timelock:       sig.Timelock,
+				CoveredFields:  sig.CoveredFields,
+				Signature:      sig.Signature,
+			})
+		}
+
+		// Insert contract.
+		return tx.Create(&dbContractRHPv2{
+			ID:         fcid,
+			Revision:   revision,
+			Signatures: signatures,
+		}).Error
+	})
```

Comment:
perhaps refactor like this?
```go
func (s *SQLContractStore) AddContract(c rhpv2.Contract) error {
 	// ... compute fcid, revision, signatures ...
 
 	dbc := &dbContractRHPv2{
 		ID:         fcid,
 		Revision:   revision,
 		Signatures: signatures,
 	}
 	return s.db.Transaction(func(tx *gorm.DB) error {
 	 	return tx.Create(dbc).Error
 	})
}
```
then it's clear that the only DB operation happening is a single `Create`.

Reply:
+1 from me, I considered commenting on this as well in a previous PR but didn't, always good to keep the amount of code inside a transaction as low as possible